File: p2_work_queues/c/recv.go
package main

import (
	"bytes"
	"log"
	"time"

	"github.com/streadway/amqp"
)

func failOnError(err error, msg string){
	if err!=nil{
		log.Fatalf("%s, %s",msg, err)
	}
}

func main() {
	conn, err := amqp.Dial("amqp://guest:guest@localhost:5672/")
	failOnError(err, "Failed to connect to RabbitMQ")
	defer conn.Close()

	ch, err := conn.Channel()
	failOnError(err, "Failed to open a channed")
	defer ch.Close()

	q, err := ch.QueueDeclare(
		"task_queue",
		true,
		false,
		false,
		false,
		nil,
	)
	failOnError(err, "Failed to declare a queue")

	err = ch.Qos(1, 0, false)
	failOnError(err, "Failed to set Qos")

	msgs, err := ch.Consume(
		q.Name,
		"",
		false,
		false,
		false,
		false,
		nil,
	)
	failOnError(err, "Failed to register a consumer")
	
	forever := make(chan bool)

	go func(){
		for d:= range msgs{
			log.Printf("Received a message : %s", d.Body)
			dot_count := bytes.Count(d.Body, []byte("."))
			t:=time.Duration(dot_count)
			time.Sleep(t*time.Second)
			log.Printf("Done")
			d.Ack(false)
		}
	}()

	log.Printf("[*] Waiting for message. To exit press CTRL+C")
	<-forever
}

File: p2_work_queues/c/recv.py
#!/usr/bin/env python
# _*_coding:utf-8_*_

"""
@Datetime :   4/16/20 2:33 PM
@Author   :   Fangyang
"""
import time
import pika


def callback(ch, method, props, body):
    print(f"[*] Received {body}")
    time.sleep(body.count(b'.'))
    print(f"[*] Done")
    ch.basic_ack(delivery_tag=method.delivery_tag)


connection = pika.BlockingConnection(pika.ConnectionParameters("localhost"))
channel = connection.channel()

channel.queue_declare(queue="task_queue", durable=True)

channel.basic_qos(prefetch_count=1)  # 保证相同的任务这个worker身上只有一个任务, 不额外分配任务
channel.basic_consume(
    queue="task_queue",
    on_message_callback=callback,
    # auto_ack=True,  # 默认为false, client处理消息时消息不丢失, client die了,由下一个client继续
)

print(f"[*] Waiting for messages. To exit press CTRL+C")
channel.start_consuming()

if __name__ == "__main__":
    pass
